kernel/stdlib.c

```diff
@@ -1,40 +1,41 @@
-#include "stdlib.h"
+#include "include/stdlib.h"
 
-void memset(void *mem, char value, size_t count) 
+void *memset(void *mem, char value, size_t count) 
 {
 	for (size_t i = 0; i < count; i++)
 	{
 		((char*)mem)[i] = value;
 	}
 }
 
-void memset_word(void *mem, uint16 value, size_t count) 
+void *memset_word(void *mem, uint16_t value, size_t count) 
 {
 	for (size_t i = 0; i < count; i++)
 	{
 		((short*)mem)[i] = value;
 	}	
 }
 
-void memcpy(void *dest, void *src, size_t count) 
+void *memcpy(void *dest, void const *src, size_t count) 
 {
+	size_t n = count / sizeof(size_t) + count % sizeof(size_t);
 	if (dest < src)
 	{
-		for (size_t i = 0; i < count; i++)
+		for (size_t i = 0; i < n; i++)
 		{
-			((char*)dest)[i] = ((char*)src)[i];
+			((size_t*)dest)[i] = ((size_t*)src)[i];

```

Comment:
you shouldn't do that without proper alignment of src and dest.

Reply:
And also with n calculated as above this will corrupt memory past the dest buffer.

Reply:
> Memory shouldn't be accessed through anything other than pointer to char without checking for proper alignment of src and dest. This is not guaranteed to be fast and it may not work at all.

Получается нужно сначала проверить выровнены ли данные. Если да то можно перемещать блоками по 4 или 8 байт. Если нет то перемещать блоками по 1 байт. Правильно ли я понял?

Reply:
Это тоже частный случай, в общем алгоритм копирующий память по словам будет ещё более сложным. Например:
```
char a[256];
char b[256];
memcpy(a + 1, b, 255);
```
-- в данном случае почти наверняка `src` и `dst` не будут выравнены одновременно.

Мне кажется, что имеет смысл начать с простой реализации копирующей побайтно и оптимизировать её по мере необходимости.